pydantic/_internal/_validators.py

```diff
@@ -330,25 +331,78 @@ def max_length_validator(x: Any, max_length: Any) -> Any:
         raise TypeError(f"Unable to apply constraint 'max_length' to supplied value {x}")
 
 
-_CONSTRAINT_TO_VALIDATOR_LOOKUP: dict[str, Callable] = {
+def _extract_decimal_digits_info(decimal: Decimal, normalized: bool) -> tuple[int, int]:
+    """Modeled after the `pydanitc-core` implementation of this function.
+
+    See https://github.com/pydantic/pydantic-core/blob/f389728432949ecceddecb1f59bb503b0998e9aa/src/validators/decimal.rs#L85.
+    Though this could be divided into two separate functions, the logic is easier to follow if we couple the computation
+    of the number of decimals and digits together.
+    """
+    normalized_decimal = decimal.normalize() if normalized else decimal
+    _, digit_tuple, exponent = normalized_decimal.as_tuple()
+
+    exponent = int(exponent)
+    digits = len(digit_tuple)
+
+    if exponent >= 0:
+        # // A positive exponent adds that many trailing zeros.
+        digits += exponent
+        decimals = 0
+    else:
+        #  If the absolute value of the negative exponent is larger than the
+        #  number of digits, then it's the same as the number of digits,
+        #  because it'll consume all the digits in digit_tuple and then
+        #  add abs(exponent) - len(digit_tuple) leading zeros after the
+        #  decimal point.
+        decimals = abs(exponent)
+        digits = max(digits, decimals)
+
+    return decimals, digits
```

Comment:
I generally agree with this change but am a bit scared by this logic. @alexmojaki would you mind double checking?

Reply:
Also intimidated by this pattern - see the `pydantic-core` ref above that does basically the exact same thing. Thanks for the review :)

Reply:
I've rewritten the docstring and added some explicit tests here to help.